docs/topics/kubernetes-developers.md

```diff
@@ -123,3 +123,48 @@ AZURE_CLOUD_NODE_MANAGER_IMG=myrepo/my-cnm:v0.0.1 \
 CLUSTER_TEMPLATE=cluster-template-external-cloud-provider.yaml \
 make create-workload-cluster
 ```
+
+## Testing clusters built from Kubernetes source
+
+[A template is provided](../../templates/cluster-template-custom-builds.yaml) that enables building clusters from custom built Kubernetes components. At this time only tilt flows have been tested. Here's an example of a `tilt-settings.json` suitable for testing from custom-built artifacts:
+
+```json
+{
+  "kustomize_substitutions": {
+    "AZURE_SUBSCRIPTION_ID_B64": "<insert base64-encoded Azure subscription ID here>",
+    "AZURE_TENANT_ID_B64": "<insert base64-encoded Azure tenant ID here>",
+    "AZURE_CLIENT_SECRET_B64": "<insert base64-encoded service principal password here>",
+    "AZURE_CLIENT_ID_B64": "<insert base64-encoded service principal ID here>",
+    "AZURE_ENVIRONMENT": "AzurePublicCloud",
+    "AZURE_SSH_PUBLIC_KEY_B64": "<insert ssh public key here>"
+  },
+  "worker-templates": {
+    "flavors": {
+      "custom-builds": {
+        "CLUSTER_NAME": "<insert cluster name here>",
+        "KUBE_BINARY_URL": "<insert URL to .tar.gz file containing Kubernetes source-built artifacts here",
+        "KUBE_APISERVER_IMAGE_URL": "<insert URL to source-built Kubernetes apiserver container image here>",
+        "KUBE_CONTROLLER_MANAGER_IMAGE_URL": "<insert URL to source-built Kubernetes controller-manager container image here>",
+        "KUBE_SCHEDULER_IMAGE_URL": "<insert URL to source-built Kubernetes scheduler container image here>",
+        "KUBE_PROXY_IMAGE_URL": "<insert URL to source-built Kubernetes kube-proxy container image here>",
+        "WORKER_MACHINE_COUNT": "3",
+        "KUBERNETES_VERSION": "1.20.5",
+        "AZURE_LOCATION": "westus2",
+        "AZURE_NODE_MACHINE_TYPE": "Standard_D2s_v3",
+        "CONTROL_PLANE_MACHINE_COUNT": "1"
+      }
+    },
+    "metadata": {}
+  }
+}
+```
+
+A few notes:
+
+- The URL referenced in a `KUBE_BINARY_URL` environment variable should be a tar'd + gzip'd file that contains files at these relative filepaths in the archive:
+  - `kubernetes/node/bin/kubelet`
+  - `kubernetes/node/bin/kubectl`
+- It's not strictly required to include a reference to every component (for example, if you just want to test a custom build of kube-proxy, you may just include the `"KUBE_PROXY_IMAGE_URL"` configuration in your `tilt-settings.json`), but we do assume that any referenced custom components passed to the cluster template are all built from the same source.
+- Specify a value of `KUBERNETES_VERSION` that shares a common minor version heritage with the source commit, if possible. If you are testing against very recent main branch upstream Kubernetes commits, the guidance is to use the most recent version of Kubernetes that your capz workflow supports.
```

Comment:
We can mention the following:
very recent main branch: use https://storage.googleapis.com/kubernetes-release/release/stable.txt
release branch: use https://storage.googleapis.com/kubernetes-release/release/stable-1.XX.txt